Consolidate file-based authorization docs (#10800)
trinodb/trino: docs/src/main/sphinx/security/file-system-access-control.rst

=========================
File-based access control
=========================

To secure access to data in your cluster, you can implement file-based access
control where access to data and operations is defined by rules declared in
manually-configured JSON files.

There are two types of file-based access control:

* **System-level access control** uses the access control plugin with a single
  JSON file that specifies authorization rules for the whole cluster.
* **Catalog-level access control** uses individual JSON files for each catalog
  for granular control over the data in that catalog, including column-level

>   authorization.

Author: How do these two types of access control interact? How is precedence handled if there's a catalog file and system file that have contradicting rules?

Author: Chatted with @electrum , the way this works is that `deny` rules always take precedence over `allow`. So if there is a rule in one file that conflicts with another, the operation will be denied according to either rule. As a result, there's no favoring of system vs catalog. Will add to this PR shortly.

=========================
File-based access control
=========================

To secure access to data in your cluster, you can implement file-based access
control where access to data and operations is defined by rules declared in
manually-configured JSON files.

> There are two types of file-based access control:

Reviewer: There are two types of access control in general (not specific to file based). Maybe we should mention this somewhere (not in this file though since it's specific to the file-based mechanisms).

Author: Currently that's described in /security/overview, which seems like the right place for it. I don't think any immediate edits need to be made there, but am open to suggestions.

Reviewer: /security/overview satisfies my ask. Resolved.

* **System-level access control** uses the access control plugin with a single
  JSON file that specifies authorization rules for the whole cluster.
* **Catalog-level access control** uses individual JSON files for each catalog
  for granular control over the data in that catalog, including column-level
  authorization.

> File-based access control process from the top to bottom of the JSON file with
> an implicit deny. If a rule matches the user then the corresponding operation
> is allowed. If no rules match, the operation is denied.

Reviewer: ```suggestion
File-based access control rules are processed in order from top to bottom of the JSON rules file. The first matching rule is applied. If no rules match, the operation is denied.
```

Note that a match doesn't have to mean "allow". e.g. `"owner": false` in a rule will remove ownership for the matched user instead of granting it.

The concept is that there is a list of "rules" (i.e. what to do - could be allow/deny/whatever) and the user information is matched against the list of rules. The first rule that matches is executed. If none match then the permission is denied.

(Feel free to use this info to reword/adjust my suggestion above).

Author: Removed this paragraph per your comment below, since it's described in the respective intros for system-level and catalog-level.

.. _system-file-based-access-control:

System-level access control files
=================================

The access control plugin allows you to specify authorization rules for the
cluster in a single JSON file.

Configuration
-------------

To use the access control plugin, add an ``etc/access-control.properties`` file
containing two required properties: ``access-control.name``, which must be set
to ``file``, and ``security.config-file``, which must be set to the location
of the config file. For example, if a config file named ``rules.json`` resides
in ``etc``, add an ``etc/access-control.properties`` with the following
contents:

.. code-block:: text

   access-control.name=file
   security.config-file=etc/rules.json

The config file is specified in JSON format. It contains rules that define which
users have access to which resources. The rules are read from top to bottom and

> the first matching rule is applied. If no rule matches, access is denied.

Reviewer: I'd recommend to keep this paragraph and remove the one at line 17 instead. The one at line 17 is introduced prematurely and it's possible in future catalog rules and system rules will have different semantics/mechanics.

Author: Done

=========================
File-based access control
=========================

To secure access to data in your cluster, you can implement file-based access
control where access to data and operations is defined by rules declared in
manually-configured JSON files.

There are two types of file-based access control:

* **System-level access control** uses the access control plugin with a single
  JSON file that specifies authorization rules for the whole cluster.
* **Catalog-level access control** uses individual JSON files for each catalog
  for granular control over the data in that catalog, including column-level
  authorization.

File-based access control always prioritizes deny outcomes. If a rule in either
a system-level or catalog-level access control file denies a particular
operation, the operation is denied even if another rule is intended to allow

> that operation.

Reviewer: I don't think this is accurate. AFAIK it's first matching rule wins.

i.e. if I have rule_allow and rule_deny (in that order) then even in the case where both rules are possible matches only rule_allow would be applied.

@kokosing can double check me on this (I looked in `io/trino/plugin/base/security/FileBasedSystemAccessControl.java`)

Reviewer 2: That is correct. We process them from to top to bottom. If none match, then access is denied. If any match then it is allowed.

I don't recall any deny rules in file based access controls. @hashhar Can you please help to me find them? If they are there, then expected behavior is to evaluate deny rules first.

Reviewer: I couldn't find any implementations either. There is SPI for it though.

Author: From my conversation with @electrum last Friday it sounded like a rule_deny anywhere would override a rule_allow anywhere. If that's wrong I'll update, but want to make sure I'm not misunderstanding.

Author: Updated the introduction, please take a look.

rule, processed from top to bottom. If no rules match, the authorization is
denied. If impersonation rules are not present but the legacy principal rules
are specified, it is assumed impersonation access control is being handled by
the principal rules, so impersonation is allowed.  If neither impersonation nor
principal rules are defined, impersonation is not allowed.

Each impersonation rule is composed of the following fields:

* ``original_user`` (optional): regex to match against the user requesting the
  impersonation. Defaults to ``.*``.
* ``original_role`` (optional): regex to match against role names of the
  requesting impersonation. Defaults to ``.*``.
* ``new_user`` (required): regex to match against the user to be
  impersonated.
* ``allow`` (optional): boolean indicating if the authentication should be
  allowed.

The following example allows the ``admin`` role, to impersonate any user, except
for ``bob``.  It also allows any user to impersonate the ``test`` user:

.. literalinclude:: user-impersonation.json
    :language: json

.. _system-file-auth-principal-rules:

Principal rules
---------------

.. warning::

>     Principal rules are deprecated and are to be removed in a future release.

Author: This note was added in release 330. Has this work been completed? Will it be completed?

Reviewer: No. Not removed yet.

Author: Got it. Removed the "will be removed later" part of this note. That sort of wording ought to be kept out of docs, because of exactly this scenario.

The fixed management user only applies to HTTP by default. To enable the fixed
user over HTTPS, set the ``management.user.https-enabled`` configuration
property.

.. _catalog-file-based-access-control:

Catalog-level access control files
==================================

You can create JSON files for individual catalogs that define authorization
rules specific to that catalog. To enable catalog-level access control files,
add a connector-specific catalog configuration property that sets the
authorization type to ``FILE`` and the ``security.config-file`` catalog
configuration property that specifies the JSON rules file.

For example, the following Iceberg catalog configuration properties use the
``rules.json`` file for catalog-level access control:

.. code-block:: properties

  iceberg.security=FILE
  security.config-file=etc/catalog/rules.json

Catalog-level access control files are supported on a per-connector basis, refer
to the connector documentation for more information.

.. note::

    These rules do not apply to system-defined tables in the
    ``information_schema`` schema.

Configure a catalog rules file
------------------------------

> Catalog-level rules JSON files are composed of three sections, each of which is
> a list of rules that are matched in order from top to bottom. The user is
> granted the privileges from the first matching rule. All regexes default to
> ``.*`` if not specified.

Reviewer: ```suggestion
The config file is specified in JSON format. It has multiple sections each of which
contains a list of rules that are matched in order from top to bottom. The user is
granted the privileges from the first matching rule. All regexes default to
``.*`` if not specified.
```

> Schema rules

Reviewer: None of these sections specify the name of the section to be used in the final JSON. We should fix this. No idea how it should look though.

Author: As a compromise, added some writing in the intro that shows the high level structure with the actual section names in code font.

^^^^^^^^^^^^

These rules govern who is considered an owner of a schema.

* ``user`` (optional): regex to match against user name.
* ``group`` (optional): regex to match against every user group the user belongs
  to.
* ``schema`` (optional): regex to match against schema name.
* ``owner`` (required): boolean indicating ownership.

Table rules
^^^^^^^^^^^

These rules govern the privileges granted on specific tables.

* ``user`` (optional): regex to match against user name.
* ``group`` (optional): regex to match against every user group the user belongs
  to.
* ``schema`` (optional): regex to match against schema name.
* ``table`` (optional): regex to match against table name.
* ``privileges`` (required): zero or more of ``SELECT``, ``INSERT``,
  ``DELETE``, ``OWNERSHIP``, ``GRANT_SELECT``.
* ``columns`` (optional): list of column constraints.
* ``filter`` (optional): boolean filter expression for the table.

> * ``filter_environment`` (optional): environment use during filter evaluation.

Reviewer: ```suggestion
* ``filter_environment`` (optional): environment used during filter evaluation.
```

Reviewer: Link from here to "Filter and mask environment".

Author: I don't think a link is needed with the restructuring/rewording below

Column constraint
~~~~~~~~~~~~~~~~~

These constraints can be used to restrict access to column data.

* ``name``: name of the column.
* ``allow`` (optional): if false, column can not be accessed.
* ``mask`` (optional): mask expression applied to column.

> * ``mask_environment`` (optional): environment use during mask evaluation.

Reviewer: ```suggestion
* ``mask_environment`` (optional): environment used during mask evaluation.
```

Reviewer: Link from here to "Filter and mask environment".

Author: I don't think the link is needed with the restructuring/wording below

Filter and mask environment
^^^^^^^^^^^^^^^^^^^^^^^^^^^

> 

Reviewer: Add some text to clarify this is for the value of `mask_environment` and `filter_environment`.

  ``read``, ``write``

For example, if you want to allow only the role ``admin`` to read and write
system information, allow ``alice`` to read system information, and deny all
other access, you can use the following rules:

.. literalinclude:: system-information-access.json
    :language: json

A fixed user can be set for management interfaces using the ``management.user``
configuration property.  When this is configured, system information rules must
still be set to authorize this user to read or write to management information.
The fixed management user only applies to HTTP by default. To enable the fixed
user over HTTPS, set the ``management.user.https-enabled`` configuration
property.

.. _catalog-file-based-access-control:

Catalog-level access control files
==================================

You can create JSON files for individual catalogs that define authorization
rules specific to that catalog. In the catalog configuration file, you then
add a configuration property that sets the authorization type to ``FILE`` and
points to this configuration file. Catalog-level access control files are
supported on a per-connector basis, refer to the connector documentation for

> more information.

Reviewer: the intro for system level control is better structured, mentioning the configs that need to be set etc.

Author: Rewrote this catalog-level intro to be more helpful and in line with the system-level intro.
